src/envs/cover.py

```diff
@@ -435,81 +437,171 @@ def action_space(self) -> Box:
+            return cross_product(subtract(c, a), subtract(b, a))
+        d1 = direction(p3, p4, p1)
+        d2 = direction(p3, p4, p2)
+        d3 = direction(p1, p2, p3)
+        d4 = direction(p1, p2, p4)
+        if ((d1 > 0 and d2 < 0) or (d1 < 0 and d2 > 0)) and \
+        ((d3 > 0 and d4 < 0) or (d3 < 0 and d4 > 0)):
+            return True
+        return False
+
+    @staticmethod
+    def overlap(l1: Tuple[float, float], r1: Tuple[float, float],
+                   l2: Tuple[float, float], r2: Tuple[float, float]):
+        # l1, r1 are top left, bottom right points of first rectangle
+        # l2, r2 are top left, bottom right points of second rectangle
+
+        if (l1[0] >= r2[0] or l2[0] >= r1[0]):  # one rect on left side of other
+            return False
+        if (r1[1] >= l2[1] or r2[1] >= l1[1]):  # one rect above the other
+            return False
+        return True
+
     def simulate(self, state: State, action: Action) -> State:
         # Since the action space is lower level, we need to write
         # a lower level simulate function.
         assert self.action_space.contains(action.arr)
+
+        # Get data needed to collision check the trajectory.
         dx, dy, dgrip = action.arr
         next_state = state.copy()
         x = state.get(self._robot, "x")
         y = state.get(self._robot, "y")
         grip = state.get(self._robot, "grip")
-        # Update robot state based on action.
+        held_block = None
+        for block in self._blocks:
+            if state.get(block, "grasp") != -1:
+                assert held_block is None
+                held_block = block
+                hx, hy = state.get(held_block, "x"), \
+                           state.get(held_block, "y")
+                hw, hh = state.get(held_block, "width"), \
+                         state.get(held_block, "height")
+
+        # Ensure neither the gripper nor the possible held block collide with
+        # another block or go below the y-axis.
+        if y + dy < 0 - self.collision_threshold:
+            return state.copy()
+        if held_block is not None:
+            if hy - hh + dy < 0 - self.collision_threshold:
+                return state.copy()
+        p1 = (x, y)
+        p2 = (x + dx, y + dy)
+        for block in self._blocks:
+            bx, by = state.get(block, "x"), state.get(block, "y")
+            bw, bh = state.get(block, "width"), state.get(block, "height")
+            ct = self.collision_threshold
+            # These segments defines a slightly smaller rectangle to prevent
+            # floating point arithmetic making us declare a false positive.
+            segments = [
+                ((bx-bw/2+ct, by-ct), (bx+bw/2-ct, by-ct)),  # top
+                ((bx+bw/2-ct, by-ct), (bx+bw/2-ct, by-bh)),  # right
+                ((bx-bw/2+ct, by-ct), (bx-bw/2+ct, by-bh))  # left
+            ]
+            # Only do the check for blocks that are not held.
+            if held_block is None or \
```

Comment:
makes sense -- how about we refactor this slightly so that we skip the whole loop over blocks if no object is held? so add before the chunk: `if held_block is not None:`

Reply:
the structure `if not X: continue` is more preferable than `if X: <indent everything>`. so the best thing would be, if i'm understanding your logic correctly....
```
if held_block is not None and block == held_block:
    continue
```
and then you don't need either this if statement, or the same one a bit below

Reply:
@tomsilver the robot can still collide with a block, so looping through all blocks in the outer loop has to happen?

Reply:
hmm I see, I didn't realize that you were checking robot-block collisions, I thought only block-block. can you point me to where the robot-block collision checks will happen?